File: src/day2/main.cpp
/**
 * Day ol - 2
 * jarviliam
 */

#include <array>
#include <fstream>
#include <iostream>
#include <vector>

void buildStrings(std::vector<std::string> &output) {
  std::ifstream file("./input.txt");
  if (file.is_open()) {
    std::string entry;
    while (std::getline(file, entry)) {
      output.push_back(entry);
    };
  };
};

int part2(std::vector<std::string> inputs) {
  int valid = 0;
  for (auto i : inputs) {
    // Retrive Minimum Val
    int delimPos = i.find("-");
    int minVal = std::stoi(i.substr(0, delimPos));
    i.erase(0, delimPos + 1);

    // Retrieve Max
    delimPos = i.find(" ");
    int maxVal = std::stoi(i.substr(0, delimPos));
    i.erase(0, delimPos + 1);

    std::cout << i << std::endl;

    // Retrieve Pass
    delimPos = i.find(":");
    char val = i[delimPos - 1];
    i.erase(0, delimPos + 1);

    //Big ol XOR
    if ((i[minVal] == val) ^ (i[maxVal] == val)) {
      valid++;
    }
  }
  return valid;
};

int main() {
  std::vector<std::string> inputs;
  buildStrings(inputs);

  int valid = 0;
  // Loop through each entry
  for (auto i : inputs) {

    // Format: 4-5 m : mmmmm
    // Retrive Minimum Val
    int delimPos = i.find("-");
    int minVal = std::stoi(i.substr(0, delimPos));
    i.erase(0, delimPos + 1);

    // Retrieve Max
    delimPos = i.find(" ");
    int maxVal = std::stoi(i.substr(0, delimPos));
    i.erase(0, delimPos + 1);

    // Retrieve Pass
    delimPos = i.find(":");
    char val = i[delimPos - 1];
    i.erase(0, delimPos + 1);

    int cnt = 0;
    for (auto ch : i) {
      if (ch == val) {
        cnt++;
      };
    };
    if (cnt >= minVal && cnt <= maxVal) {
      valid++;
    };
  };
  std::cout << "Answers : " << valid << std::endl;
  std::cout << "Input Size : " << inputs.size() << std::endl;

  std:: cout << part2(inputs) << std::endl;
};

File: go/day2/main.go
package main

import (
	"bufio"
	"log"
	"os"
	"regexp"
	"strconv"
)

func countRune(str string, ch rune) int {
	cnt := 0
	for _, i := range str {
		if i == ch {
			cnt++
		}
	}
	return cnt
}

func part1(inputs []string) int {
	//God Bless Regexp
	// Format: 4-5 m : mmmmm
	reg, _ := regexp.Compile("(\\d+)-(\\d+) ([a-z]): ([a-z]+)")

	valid := 0
	for _, passport := range inputs {
		match := reg.FindStringSubmatch(passport)

		minVal, _ := strconv.Atoi(match[1])
		maxVal, _ := strconv.Atoi(match[2])
		target := rune(match[3][0])
		cnt := countRune(match[4], target)
		if cnt >= minVal && cnt <= maxVal {
			valid++
		}
	}
	return valid
}

func part2(inputs []string) int {
	reg, _ := regexp.Compile("(\\d+)-(\\d+) ([a-z]): ([a-z]+)")
	valid := 0
	for _, passport := range inputs {
		match := reg.FindStringSubmatch(passport)

		minVal, _ := strconv.Atoi(match[1])
		maxVal, _ := strconv.Atoi(match[2])
		target := match[3][0]
		pass := match[4]

		minVal -= 1
		maxVal -= 1
		if maxVal >= len(pass) {
			continue
		}
        //Damn you go for not having XOR
		if ((pass[minVal] == target) || (pass[maxVal] == target)) && !((pass[maxVal] == target) && (pass[minVal] == target)) {
			valid++
		}
	}
	return valid
}

func loadLines(path string) ([]string, error) {
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var inputs []string
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		inputs = append(inputs, scanner.Text())
	}
	return inputs, scanner.Err()
}

func main() {
	inputs, _ := loadLines("./input.txt")
	out1 := part1(inputs)
	out2 := part2(inputs)
	log.Printf("Answer 1 : %d\n", out1)
	log.Printf("Answer 2 : %d\n", out2)
}
